Engineering/Mobile/Kotlin coroutine.md

```diff
@@ -0,0 +1,170 @@
+date: 2022-09-06
+---
+
+Start from Kotlin 1.3, JetBrain introduce coroutine a light-weight threads. This article will focus on basic concept of coroutine and how you can use it in Kotlin.
+
+## Coroutine overview
+
+Coroutine, an instance of suspendable computation that takes a block of code and run it on concurrently with the rest of the program. Concept of coroutine is familar with thread but coroutine does not bound to any particular thread. It can suspend its execution in this thread and resume it on another one.
+
+## Structured concurrency
+
+Before we dive into how we can use Coroutine in Kotlin, let me first introduce Structured concurrency, the principle of coroutine.
+
+>Due to [wikipedia](https://en.wikipedia.org/wiki/Structured_concurrency), the core concept is to encapsulate the concurrent threads of execution, so that we can control the flow construct with clear entry and exit points. Also, ensure all children must be completed before exit. A scenario that proves this pattern:
+
+Let's say you want to make breakfast with fried eggs and bread. I assume that we have a total of 3 tasks that need to be done:
+- Frying the eggs
+- Put our bread into the toaster
+- Bring everything on a disk
+
+Structured concurrency present that all sub-tasks shall complete before the completion of their parent task(s). No sub-tasks can outlive its parent task(s). The principle also ensures that any errors that happen in its child are properly reported and are never leaked.
+
+## Kotlin coroutines
+
+You can run the following code block on [Kotlin playground](https://play.kotlinlang.org/).
+
+```kotlin
+import kotlinx.coroutines.*
+
+fun main() = runBlocking {
+    // Parent task
+    val parentJob = launch {
+
+        // Child task
+		launch {
+            var count = 1
+            while (count <= 5) {
+                println("Count: $count")
+                delay(100)
+                count++
+            }
+        }
+    }
+
+    delay(250)
+    println("Cancelling parent job")
+    parentJob.cancel()
+}
+
+--------------- Output ----------------
+Count: 1
+Count: 2
+Count: 3
+Cancelling parent job
+```
+
+Let's me explain some basic functions:
+- `runBlocking {...}` is a coroutine builder. It is designed to bridge the non-coroutine code of a regular `fun main()` with all coroutine code inside `runBlocking` lambda.
+- `launch {...}` is also a coroutine builder. Use this when you want to launch a new coroutine concurrently with the rest of the code, that can continues to work independently.
+- `delay()` is a suspend function, this work almost the same as `sleep()` function from Java. But because this is a suspend function, it does not block the current thread and allows other coroutines code to run and use the current thread.
+
+As we know [[structured concurrency]] principle, all sub-tasks cannot outlive their parent task. The example above shows exactly this when the child's task can only print count 3 times before its parent canceled. 
```

Comment:
I don't think we had  `[[structured concurrency]]` note. Did you mean to refer it to somewhere else?

Reply:
Sorry, I missed this one. Just a highlight worlds